README.md
## Changelog
- 2020-09-06 v2.2 - [LINK](CHANGELOG.md)
- 2020-08-21 v2.1 - [LINK](CHANGELOG.md)

# zdownloader
Zdownloader is a lightweight download manager running from command line.  
Zdownloader was written in C++ and supports downloading files by HTTPS from services such as:
- ZippyShare,
- Google Drive,
- MegaUp(megaup.net).

![screenshot](images/zd_running_pub.png)

## Features
- Can download multiple files at a time.
- Supports segmented downloading.
- It can resume interrupted downloads.
- Can check multiple links at a time.
- It can reconnect after specific time if download speed is too low.
- It can unpack archive files after download. Supports: rar, 7z and zip.
- Supports network proxies.
- Allows you to download google drive files even if the daily limit of download has exceeded using google drive REST API v3.

## Requirements
#### Linux x86-64  
`ldd zdownloader`:
- linux-vdso.so.1
- libbotan-2.so.13
- libQt5Network.so.5 
- libQt5Core.so.5
- libQt5Script.so.5
- libpthread.so.0
- libstdc++.so.6
- libgcc_s.so.1
- libc.so.6
- libz.so.1
- libdl.so.2
- libm.so.6
- /lib64/ld-linux-x86-64.so.2

#### General
- Linux kernel >= 3.2.0
- glibc(libc6) >= 2.25
- OpenSSL >= 1.1.0
- Qt5 >= 5.13.0  
I have attached precompiled Qt libs in lib directories (license: GNU LGPL version 3, Qt source code: https://code.qt.io)
- Botan >= 2.14 - needed for JWT sign in gdrive api  
I have attached precompiled Botan library in lib directory (https://botan.randombit.net)
- for unpack rar files: unrar >= 5.50
- for unpack 7z, zip files: 7z >= 16.02

#### Windows
- you probably need Visual C++ 2017 redistributables x64: https://aka.ms/vs/16/release/vc_redist.x64.exe
- I have attached precompiled Qt libs in win7_8_10-x64 directory. License: GNU LGPL version 3, Qt source code: https://code.qt.io.
- I have attached precompiled OpenSSL libs in win7_8_10-x64 directory - https://www.openssl.org
- for unpack rar files: WinRar >= 5.50 - https://www.rarlab.com/download.htm
- for unpack 7z, zip files: 7-zip >= 16.02 - https://www.7-zip.org/download.html

## Build instructions
- [BUILD HOWTO](BUILD_README.md)

## Usage
**links.txt** file:
- create links.txt file in zdownloader directory,
- you can add URLs to links.txt file: one line -> one URL,
- at the beginning zdownloader checks and sorts the links,
- checked and sorted links are saved to download_list.txt file,
- you can add '--' chars to append links as one group,
- you can add encoded links,
- you can put line comments - just add hash(#) character before comment.
- from zdownloader v2.0 you can use(in links.txt file) any of these gdrive link formats (and similar):  
https://drive.google.com/file/d/0B1MVW1mFO2zmZHVRWEQ3Rkc3SVE  
https://drive.google.com/file/d/0B1MVW1mFO2zmZHVRWEQ3Rkc3SVE/view  
https://drive.google.com/uc?id=0B1MVW1mFO2zmZHVRWEQ3Rkc3SVE  
https://drive.google.com/uc?id=0B1MVW1mFO2zmZHVRWEQ3Rkc3SVE&export=download

### Example of links.txt file:
```
# this is comment
https://www73.zippyshare.com/v/0deWxCol/file.html
https%3A%2F%2Fwww73.zippyshare.com%2Fv%2F0deWxCol%2Ffile.html
--
https://drive.google.com/file/d/2M4RhNINIETzt-a6tbkdwLTBS9CAqM-WD
--
https://megaup.net/3Pwaa/something.rar #comment: example link
```

### download_list.txt file:
- download_list.txt file is created after links checks,
- download list has format:
```
file_name URL group_id file_size
```
Four fields separated by tab character.  
Group_id contains group number - is used for unpack archives.  
- You can edit(eg. change order) download_list.txt when zdownloader is not running,
- If remote file does not exists then zdownloader skips that file.

### finished_downloads_\*.txt files:
- contains timestamp, download status, file name and URL,
- download_status can be:  
F (download finished with success),  
X (skipped download, file already exists on disk),  
R (skipped download, remote file does not exists)

### How gdrive download limit bypass works?
- Zdownloader can download google drive files even if the daily limit of download has exceeded. For that zdownloader can use google drive REST API v3.
- Before you start see how to enable and create service account: [LINK](HOWTO_gdrive_service_account.md)
1. The Google Drive API provides an option for 3rd party applications to store their own data on Google Drive, but sandboxed from the user's main file storage area. Hence zdownloader cannot read, modify or delete your Google Drive data. A user's space allocation (default 15GB) is shared between all applications that store data in their Google Account - Drive, Gmail, Google Photos and 3rd party apps like zdownloader.
2. At first step zdownloader list all files on gdrive.
3. Next zdownloader deletes all found files.
4. At last step the application copies the file to your gdrive and download it. 

When using gdrive api you can download one gdrive file at a time(but you can use many segments and other services like zippyshare).


### Unpack
Default value for unrar_binary setting:  
- Linux: `unrar`, Windows: `C:\Program Files\WinRAR\UnRar.exe`

Default value for 7z_binary setting:  
- Linux: `7z`, Windows: `C:\Program Files\7-zip\7z.exe`

Executing parameters:
- rar: `unrar x -parchive_pwd -ai -y -c- -o+ archive_filename archive_dir/`
- 7z: `7z x -parchive_pwd -aoa -y -oarchive_dir/ archive_filename`  
where `archive_dir` is file name without extension.

While the unpack is running, `unpack_queue.txt` file is created in the current directory.

### Start and stop
You can run zdownloader by  
`./start_zd.sh`  
from zdownloader directory.

To stop zdownloader press Ctrl-C or send SIGINT or SIGTERM signal.

### Running zdownloader on MacOS
- Unpack `macos_x64.tgz` and copy `zdownloader.app` to user's home directory.
- Edit `zdownloader.ini` in `zdownloader.app/Contents/MacOS/zdownloader.ini`
- Edit `links.txt` in `zdownloader.app/Contents/MacOS/links.txt`
- Run Terminal and write commands:
```
cd zdownloader.app/Contents/MacOS/
./zdownloader
```

#### Settings are stored in zdownloader.ini file:
Hint: in `zdownloader.ini` file you cannot use tilde(~) character as alias to user's home directory.

```ini
; UTF-8 encoding

[Common]
; download_directory
; if directory contains spaces then enclose it in quotation marks
; set empty or '.' for current directory
; on Windows use slash or double backslash as dir separators
download_directory=/media/downloads
user_agent_for_https_connections="Mozilla/5.0 (Windows NT 6.1; rv:60.0) Gecko/20100101 Firefox/60.0"
; shutdown_system_on_finish
; uses 'sudo shutdown --no-wall -P +2' command on Linux or 'shutdown /s /t 120' command on Windows
shutdown_system_on_finish=false
; max_seconds_before_reset_inactive_connection
; integer value
max_seconds_before_reset_inactive_connection=45
; download_lists_directory
; program will read file 'download_list.txt' from download_lists_directory
; program will save file 'finished_downloads_*.txt' in download_lists_directory
download_lists_directory=
; links_file_directory
; program will read file 'links.txt' from links_file_directory
links_file_directory=

[LinkChecker]
max_parallel_link_checks=5
; max_seconds_before_check_link_again_on_error
; integer value
max_seconds_before_check_link_again_on_error=10

[Downloader]
max_parallel_file_downloads=1
max_segments_per_file=2
; min_segment_size_mb
; integer value. Do not change min_segment_size_mb when you have unfinished downloads
min_segment_size_mb=3
; segment_network_read_buffer_kb
; integer value. Download read buffer(in RAM) size per one segment in KBytes.
; Application will try to stop reading from the network once this buffer is full, thus causing the download to throttle down as well.
; This can happen when your system is not capable to download with full speed.
segment_network_read_buffer_kb=2048
; segment_file_write_buffer_kb
; integer value. File write buffer(in RAM) size for one segment in KBytes.
segment_file_write_buffer_kb=1024
; min_seconds_before_next_download, max_seconds_before_next_download, next_download_increase_step_msecs
; next_download_increase_reset_after_seconds
; integer values. Applies to per file download (not per segment)
min_seconds_before_next_download=0
max_seconds_before_next_download=7
next_download_increase_step_msecs=250
; next_download_increase_reset_after_seconds
; integer value. Resets next file download delay to 0 seconds after X seconds
next_download_increase_reset_after_seconds=8
; progress_check_interval_seconds
; integer value. Show 'Progress:...' message every X seconds
progress_check_interval_seconds=10
; speed_check_interval_seconds
; integer value. Show actual speed and average speed every X seconds
speed_check_interval_seconds=8
; user_reconnect_interval_minutes
; integer value. Reconnects all connections every X minutes if avg speed is less than min_avg_speed_before_reconnect_mbps
user_reconnect_interval_minutes=4
; min_avg_speed_before_reconnect_mbps
; integer value
min_avg_speed_before_reconnect_mbps=13

[OutputMessages]
; log_to_file_enabled
; enable or disable applog_*.log files. Applog contains all output messages.
log_to_file_enabled=true
log_directory=
; timestamp_format
; format details: https://doc.qt.io/archives/qt-5.13/qdatetime.html#toString
timestamp_format="yyyyMMdd.hh:mm:ss"
; log_max_rotate_kb
; integer value in KBytes
log_max_rotate_kb=4096
log_max_archive_files=3

[FinishedDownloads]
; log_to_file_enabled
; enable or disable log finished items to finished_downloads_*.txt files.
log_to_file_enabled=true
; timestamp_format
; format details: https://doc.qt.io/archives/qt-5.13/qdatetime.html#toString
timestamp_format="yyyyMMdd.hh:mm:ss"
; log_max_rotate_kb
; integer value in KBytes
log_max_rotate_kb=4096
log_max_archive_files=3

[Unpack]
; unpack_after_download
; enable/disable unpack archive files after download all item in group
unpack_after_download=true
; pause_download_before_unpack
pause_download_before_unpack=false
; delete_archive_files_after_unpack
; true - remove archive files after successful unpack
delete_archive_files_after_unpack=true
; unrar_binary
; path to unrar/rar binary, eg. "/usr/bin/unrar". Leave empty for default value.
unrar_binary=
; 7z_binary
; path to 7zip binary, eg. "/usr/bin/7z". Leave empty for default value.
7z_binary=

[UnpackPasswords]
; set passwords for unpack - update size if you add/remove password
; you can set passwords here even if archive file is not locked by password
1\pass=mypass
;2\pass=mypass2
;3\pass=mypass3
size=1

[UnpackLog]
; log_to_file_enabled
; enable or disable log unpack messages to unpacklog_*.txt files.
log_to_file_enabled=true
; log_directory
; program will save 'unpacklog_*.txt' files in download_lists_directory
; if directory contains spaces then enclose it in quotation marks
; set empty or '.' for current directory
; on Windows use slash or double backslash as dir separators
log_directory=
; timestamp_format
; format details: https://doc.qt.io/archives/qt-5.13/qdatetime.html#toString
timestamp_format="yyyyMMdd.hh:mm:ss"
; log_max_rotate_kb
; integer value in KBytes
log_max_rotate_kb=512
log_max_archive_files=2

[NetworkProxy]
; enable_network_proxy
; set true if you want forward all network activity through proxy server
enable_network_proxy=false
; proxy_type
; available values: socks5 or http. Note: http proxy must support SSL.
proxy_type=socks5
; hostname
; you can set either IP address or hostname
hostname=127.0.0.1
; port
; integer value
port=8889
; username/password authentication
username=
password=

[GDrive]
; use_gdrive_api
; set use_gdrive_api=true to bypass gdrive download limit
; requirements: google account, enabled gdrive api, google service account
use_gdrive_api=false
; google_service_account_json_file
; path to json file with google service account credentials,
; most important fields in json file: client_email and private_key
google_service_account_json_file=./gdrive.json
```

Zdownloader can power off system on download finish. For that zdownloader uses  
- for Linux: `sudo shutdown --no-wall -P +2` command which should execute without any passwords,
- for Windows: `shutdown /s /t 120` command.

#### applog_\*.log files:
Zdownloader can save everything from standard output to rotated log files:  
`applog_1.log ...`

#### \*.zdm files:
When the download is interrupted, the program saves segment offsets to \*.zdm files in `download_directory`.  
The zdm file also contains the number of remaining bytes to download.

## All contributions are welcome
#### Help and contributions can be any of the following:
- star the project
- new feature or bug fixing via a pull requests
- report issues to the project issues page
- other ideas

##### If you like this project please give it a star ⭐. Thanks!

## Download:
- for Linux x86-64: [DOWNLOAD](https://github.com/sin2000/zdownloader/releases/download/2.2/linux-x86-64.zip)
- for Linux ARM gnueabihf (eg. Raspberry Pi >= 3): [DOWNLOAD](https://github.com/sin2000/zdownloader/releases/download/2.2/arm-linux-gnueabihf.zip)
- for Linux ARM aarch64 (eg. Odroid C2): [DOWNLOAD](https://github.com/sin2000/zdownloader/releases/download/2.2/aarch64-linux-gnu.zip)
- for Windows 7, 8, 10 x64: [DOWNLOAD](https://github.com/sin2000/zdownloader/releases/download/2.2/win7_8_10-x64.zip)
- for Windows 7, 8, 10 x86(32bit): [DOWNLOAD](https://github.com/sin2000/zdownloader/releases/download/2.2/win7_8_10-x86.zip)
- for MacOS x64 (tested on Mojave): [DOWNLOAD](https://github.com/sin2000/zdownloader/releases/download/2.2/macos_x64.tgz)


CHANGELOG.md
# Changelog
### Changes in v2.2 (2020-09-06):
- added: more universal zippyshare link extraction
- fixed: skipping non existing files faster
- added: skipping files with not found http status(404)
- hide Qt warning: 'QNetworkReplyImplPrivate::error:Internal problem, this method must only be called once'
- minor code optimize

### Changes in v2.1 (2020-08-21):
- fix for zippyshare
- updated precompiled Qt libraries to 5.15

### Changes in v2.0 (2020-04-24):
- added: javascript engine for extracting zippyshare links (https://github.com/sin2000/zdownloader/issues/8)
- added: after checking all links, the program will change the links.txt file name to old_links.txt
- added: statistics of download remaining items. Statistics are shown every 60 seconds.
- added: new UnpackLog section in zdownloader.ini file. Most of unpack messages will be now logged under unpacklog_\*.txt files instead of applog_\*.txt
- added: on Linux added SIGHUP signal handler which results in the correct closing of the application under 'screen' like apps at system shutdown
- added: better file grouping in download_list.txt (https://github.com/sin2000/zdownloader/issues/7)
- added: recognition of probably all gdrive file link types
- minor fixes

### Changes in v1.9 (2020-04-08):
- added: bypass gdrive download limit by using google drive api (see [readme](https://github.com/sin2000/zdownloader/blob/master/README.md) and [service account howto](https://github.com/sin2000/zdownloader/blob/master/HOWTO_gdrive_service_account.md))
- fixed: the program will crash when user stops it(Ctrl-C/SIGINT) immediately after starting it(after 'Loading state')

### Changes in v1.8 (2020-03-23):
- changed: zippyshare link extractor algorithm
- changed: skipping links in link checker when link is invalid(for example: is already on download list)
- added: micro scheduler - see zdownloader.ini file
- fixed: in very rarely cases download file can be corrupted when server returns other content than file bytes

### Changes in v1.7 (2020-02-17):
- changed: zippyshare link extractor algorithm

### Changes in v1.6 (2020-02-07):
- added: network proxy support (socks5, http)

### Changes in v1.5 (2020-02-07):
- fixed: sometimes crash after download (regression from version v1.2 - use pointer after free). This is second one since v1.2(sic!) but should be last one

### Changes in v1.4 (2020-02-06):
- fixed: sometimes crash after download (regression from version v1.2 - use pointer after free)
- change: do not shedule system shutdown in certain situation when user/system stops application

### Changes in v1.3 (2020-02-04):
- changed: zippyshare link extractor algorithm

### Changes in v1.2 (2020-02-02):
- added: unpack rar, 7z and zip files after download with multi-password support
- added finished_downloads_\*.txt files that contains only finished or skipped downloads
- download_list.txt now have only pending downloads items and group_id in last column
- added ETA(estimated time of download end) time to every progress message
- skipping dead links and download files that does not exists on remote anymore or files that already exists locally
- better output messages

### Changes in v1.1 (2019-12-26):
- fixed null pointer access when remote file does not exists
- fixed 100% CPU usage after more than 100 links checks (workaround for cleanConnections every 2mins by QNetworkAccessManager which causes CPU spikes)
- added ability to connect more than 6 times to one host (workaround for QNetworkAccessManager 6 connections limit)
- added segmented downloading
- added more settings to zdownloader.ini file
- now you can put line comments in links.txt file - just write hash(#) character before comment
- better output messages
- added ability to download small files by google drive
- now google drive links like https://drive.google.com/open?id=7B3Uon3LO_we3zF8NVaIEGNJeNAR are recognized as file downloads
- updated precompiled Qt library to 5.14.0


src/zdownloader/src/cmd_runner.cpp
#include "cmd_runner.h"
#include "app_settings.h"
#include "unpack.h"
#include "fn_scheduler.h"
#include <meta_object_ext.h>
#include <QCoreApplication>
#include <QTimer>
#include <QDir>
#include <QProcess>
#include <csignal>
#include <qdebugex.h>

cmd_runner::cmd_runner(app_settings * app_prefs)
  :app_set(app_prefs),
   zdown_net(new zdownloader_net_lib(false, this)),
   init_finished(false),
   shutdown_system_on_finish(false),
   all_downloads_finished(false),
   unpack_mod(new unpack(this)),
   start_download_sched(nullptr),
   stop_download_sched(nullptr)
{
  connect(zdown_net, &zdownloader_net_lib::init_finished, this, &cmd_runner::zdown_net_init_finished);
  connect(unpack_mod, &unpack::unpack_queue_finished, this, &cmd_runner::unpack_queue_finished, Qt::QueuedConnection);
  zdown_net->start();
}

cmd_runner::~cmd_runner()
{
  delete start_download_sched;
  delete stop_download_sched;
  delete app_set;

  qDebug() << "destructor cmd_runner";
}

void cmd_runner::unix_signal_received(int signal)
{
#ifndef SYSTEM_WINDOWS
  if(init_finished == false)
    return;

  switch(signal)
  {
    case SIGUSR1:
      reload_settings();
      break;
    case SIGUSR2:
      break;
    default:
      exit_app(false);
  }
#else
  Q_UNUSED(signal)
#endif
}

void cmd_runner::win_signal_received()
{
  if(init_finished == false)
    return;

  qDebug() << "win signal received";
  exit_app(false);
}

void cmd_runner::zdown_net_init_finished()
{
  init_finished = true;

  if(apply_settings())
  {
    connect(zdown_net->get_dl_mgr(), &download_manager::load_state_finished, this, &cmd_runner::load_state_finished, Qt::QueuedConnection);
    connect(zdown_net->get_dl_mgr(), &download_manager::download_finished, this, &cmd_runner::download_finished, Qt::QueuedConnection);
    connect(zdown_net->get_dl_mgr(), &download_manager::last_item_in_group_finished, this, &cmd_runner::downloader_last_item_in_group_finished, Qt::QueuedConnection);

    meta_object_ext::invoke_async(this, &cmd_runner::load_state);
  }
  else
  {
    exit_app(false);
  }
}

void cmd_runner::reload_settings()
{
  qDebug() << "RELOADING SETTINGS...";

  if(app_set)
  {
    app_set->load();
    apply_settings();
  }
}

bool cmd_runner::apply_settings()
{
  qDebug() << "SETTINGS BEGIN";
  const QString general_val_msg = app_set->general.validate();
  if(general_val_msg.isEmpty() == false)
  {
    qDebug() << "settings error:";
    qDebug() << general_val_msg;
    return false;
  }

  qDebug() << "Common:"
           << "\n" << "download_directory:" << app_set->general.download_directory
           << "\n" << "user_agent_for_https_connections:" << app_set->general.user_agent_for_https_connections
           << "\n" << "shutdown_system_on_finish:" << app_set->general.shutdown_system_on_finish
           << "\n" << "max_seconds_before_reset_inactive_connection:" << app_set->general.max_seconds_before_reset_inactive_connection
           << "\n" << "download_lists_directory:" << app_set->general.download_lists_directory
           << "\n" << "links_file_directory:" << app_set->general.links_file_directory;

  zdown_net->get_dl_mgr()->invoke_set_download_directory(app_set->general.download_directory);
  zdown_net->get_dl_mgr()->invoke_set_user_agent_for_https_connections(app_set->general.user_agent_for_https_connections);
  zdown_net->get_dl_mgr()->invoke_set_max_seconds_before_abort_inactive_connection(app_set->general.max_seconds_before_reset_inactive_connection);
  zdown_net->get_dl_mgr()->invoke_set_download_lists_dir(app_set->general.download_lists_directory);
  zdown_net->get_dl_mgr()->invoke_set_links_file_dir(app_set->general.links_file_directory);

  const QString link_chk_val_msg = app_set->link_checker.validate();
  if(link_chk_val_msg.isEmpty() == false)
  {
    qDebug() << "settings error:";
    qDebug() << link_chk_val_msg;
    return false;
  }

  qDebug() << "LinkChecker:"
           << "\n" << "max_parallel_link_checks:" << app_set->link_checker.max_parallel_link_checks
           << "\n" << "max_seconds_before_check_link_again_on_error:" << app_set->link_checker.max_seconds_before_check_link_again_on_error;

  zdown_net->get_dl_mgr()->invoke_set_max_parallel_link_checks(app_set->link_checker.max_parallel_link_checks);
  zdown_net->get_dl_mgr()->invoke_set_max_seconds_before_check_link_again_on_error(app_set->link_checker.max_seconds_before_check_link_again_on_error);

  const QString downloader_val_msg = app_set->downloader.validate();
  if(downloader_val_msg.isEmpty() == false)
  {
    qDebug() << "settings error:";
    qDebug() << downloader_val_msg;
    return false;
  }

  qDebug() << "Downloader:"
           << "\n" << "max_parallel_file_downloads:" << app_set->downloader.max_parallel_file_downloads
           << "\n" << "max_segments_per_file:" << app_set->downloader.max_segments_per_file
           << "\n" << "min_segment_size_mb:" << app_set->downloader.min_segment_size_mb
           << "\n" << "segment_network_read_buffer_kb:" << app_set->downloader.segment_network_read_buffer_kb
           << "\n" << "segment_file_write_buffer_kb:" << app_set->downloader.segment_file_write_buffer_kb
           << "\n" << "min_seconds_before_next_download:" << app_set->downloader.min_seconds_before_next_download
           << "\n" << "max_seconds_before_next_download:" << app_set->downloader.max_seconds_before_next_download
           << "\n" << "next_download_increase_step_msecs:" << app_set->downloader.next_download_increase_step_msecs
           << "\n" << "next_download_increase_reset_after_seconds:" << app_set->downloader.next_download_increase_reset_after_seconds
           << "\n" << "progress_check_interval_seconds:" << app_set->downloader.progress_check_interval_seconds
           << "\n" << "speed_check_interval_seconds:" << app_set->downloader.speed_check_interval_seconds
           << "\n" << "user_reconnect_interval_minutes:" << app_set->downloader.user_reconnect_interval_minutes
           << "\n" << "min_avg_speed_before_reconnect_mbps:" << app_set->downloader.min_avg_speed_before_reconnect_mbps;


  zdown_net->get_dl_mgr()->invoke_set_max_parallel_connections(app_set->downloader.max_parallel_file_downloads);
  zdown_net->get_dl_mgr()->invoke_set_max_segments_per_file(app_set->downloader.max_segments_per_file);
  zdown_net->get_dl_mgr()->invoke_set_min_segment_size_mb(app_set->downloader.min_segment_size_mb);
  zdown_net->get_dl_mgr()->invoke_set_segment_network_read_buffer_kb(app_set->downloader.segment_network_read_buffer_kb);
  zdown_net->get_dl_mgr()->invoke_set_segment_file_write_buffer_kb(app_set->downloader.segment_file_write_buffer_kb);
  zdown_net->get_dl_mgr()->invoke_set_next_download_timer_params(app_set->downloader.min_seconds_before_next_download,
                                                                          app_set->downloader.max_seconds_before_next_download,
                                                                          app_set->downloader.next_download_increase_step_msecs,
                                                                          app_set->downloader.next_download_increase_reset_after_seconds);
  zdown_net->get_dl_mgr()->invoke_set_progress_check_interval(app_set->downloader.progress_check_interval_seconds);
  zdown_net->get_dl_mgr()->invoke_set_speed_check_interval(app_set->downloader.speed_check_interval_seconds);
  zdown_net->get_dl_mgr()->invoke_set_user_reconnect_interval(app_set->downloader.user_reconnect_interval_minutes);
  zdown_net->get_dl_mgr()->invoke_set_min_avg_speed_before_reconnect(app_set->downloader.min_avg_speed_before_reconnect_mbps);

  const QString out_msgs_val_msg = app_set->out_msgs.validate();
  if(out_msgs_val_msg.isEmpty() == false)
  {
    qDebug() << "OutputMessages settings error:";
    qDebug() << out_msgs_val_msg;
    return false;
  }

  qDebug() << "OutputMessages:"
           << "\n" << "log_to_file_enabled:" << app_set->out_msgs.log_to_file_enabled
           << "\n" << "log_directory:" << app_set->out_msgs.log_directory
           << "\n" << "timestamp_format:" << app_set->out_msgs.timestamp_format
           << "\n" << "log_max_rotate_kb:" << app_set->out_msgs.log_max_rotate_kb
           << "\n" << "log_max_archive_files:" << app_set->out_msgs.log_max_archive_files;

  const QString fin_dls_val_msg = app_set->fin_dls.validate();
  if(fin_dls_val_msg.isEmpty() == false)
  {
    qDebug() << "FinishedDownloads settings error:";
    qDebug() << fin_dls_val_msg;
    return false;
  }

  qDebug() << "FinishedDownloads:"
           << "\n" << "log_to_file_enabled:" << app_set->fin_dls.log_to_file_enabled
           << "\n" << "timestamp_format:" << app_set->fin_dls.timestamp_format
           << "\n" << "log_max_rotate_kb:" << app_set->fin_dls.log_max_rotate_kb
           << "\n" << "log_max_archive_files:" << app_set->fin_dls.log_max_archive_files;

  qDebug() << "Unpack:"
           << "\n" << "unpack_after_download:" << app_set->unpack.unpack_after_download
           << "\n" << "pause_download_before_unpack:" << app_set->unpack.pause_download_before_unpack
           << "\n" << "delete_archive_files_after_unpack:" << app_set->unpack.delete_archive_files_after_unpack
           << "\n" << "unrar_binary:" << app_set->unpack.unrar_binary
           << "\n" << "7z_binary:" << app_set->unpack.seven_zip_binary;

  const QString unpack_pwds_val_msg = app_set->unpack.validate();
  if(unpack_pwds_val_msg.isEmpty() == false)
  {
    qDebug() << "UnpackPasswords settings error:";
    qDebug() << unpack_pwds_val_msg;
    return false;
  }

  qDebug() << "UnpackPasswords:"
           << "\n" << app_set->unpack.passwords;

  if(apply_unpacklog_settings() == false)
    return false;

  qDebug() << "NetworkProxy:"
           << "\n" << "enable_network_proxy:" << app_set->net_proxy.enable_network_proxy
           << "\n" << "proxy_type:" << ((app_set->net_proxy.proxy_type == QNetworkProxy::Socks5Proxy) ? QString("socks5") : QString("http"))
           << "\n" << "hostname:" << app_set->net_proxy.hostname
           << "\n" << "port:" << app_set->net_proxy.port
           << "\n" << "username:" << app_set->net_proxy.username
           << "\n" << "password:" << app_set->net_proxy.password;

  // ======
  if(apply_scheduler_settings() == false)
    return false;

  if(apply_gdrive_settings() == false)
    return false;

  qDebug() << "SETTINGS END";

  shutdown_system_on_finish = app_set->general.shutdown_system_on_finish;
  unpack_mod->set_settings(app_set->unpack);
  unpack_mod->set_working_directory(app_set->general.download_directory);
  set_application_proxy(app_set->net_proxy);

  return true;
}

bool cmd_runner::apply_unpacklog_settings()
{
  const QString val_msg = app_set->unpacklog.validate();
  if(val_msg.isEmpty() == false)
  {
    qDebug() << "UnpackLog settings error:";
    qDebug() << val_msg;
    return false;
  }
  if(app_set->unpacklog.log_to_file_enabled)
  {
    qDebug() << "UnpackLog:"
             << "\n" << "log_to_file_enabled:" << app_set->unpacklog.log_to_file_enabled
             << "\n" << "log_directory:" << app_set->unpacklog.log_directory
             << "\n" << "timestamp_format:" << app_set->unpacklog.timestamp_format
             << "\n" << "log_max_rotate_kb:" << app_set->unpacklog.log_max_rotate_kb
             << "\n" << "log_max_archive_files:" << app_set->unpacklog.log_max_archive_files;

    unpack_mod->set_log_settings(app_set->unpacklog);
  }

  return true;
}

bool cmd_runner::apply_scheduler_settings()
{
  const QString sched_val_msg = app_set->scheduler.validate();
  if(sched_val_msg.isEmpty() == false)
  {
    qDebug() << "Scheduler settings error:";
    qDebug() << sched_val_msg;
    return false;
  }
  if(app_set->scheduler.scheduler_enabled)
  {
    qDebug() << "Scheduler:"
             << "\n" << "start_download_at:" << app_set->scheduler.start_download_at.time()
             << "\n" << "stop_download_at:" << app_set->scheduler.stop_download_at.time();

    if(start_download_sched == nullptr)
      start_download_sched = new fn_scheduler;
    if(stop_download_sched == nullptr)
      stop_download_sched = new fn_scheduler;

    start_download_sched->set_activation_time(app_set->scheduler.start_download_at.time());
    stop_download_sched->set_activation_time(app_set->scheduler.stop_download_at.time());

    connect(start_download_sched, &fn_scheduler::activated, this, &cmd_runner::start_downloads, Qt::UniqueConnection);
    connect(stop_download_sched, &fn_scheduler::activated, this, &cmd_runner::scheduler_stop_triggered, Qt::UniqueConnection);
  }

  return true;
}

bool cmd_runner::apply_gdrive_settings()
{
  const QString val_msg = app_set->gdrive.validate();
  if(val_msg.isEmpty() == false)
  {
    qDebug() << "GDrive settings error:";
    qDebug() << val_msg;
    return false;
  }
  if(app_set->gdrive.use_gdrive_api)
  {
    qDebug() << "GDrive:"
             << "\n" << "use_gdrive_api:" << app_set->gdrive.use_gdrive_api
             << "\n" << "google_service_account_json_file:" << app_set->gdrive.google_service_account_json_file;


    zdown_net->get_dl_mgr()->invoke_set_gdrive_settings(app_set->gdrive);
  }

  return true;
}

void cmd_runner::set_application_proxy(const net_proxy_settings & proxy_prefs)
{
#ifdef DEBUG_PROXY
  QNetworkProxy proxy(QNetworkProxy::Socks5Proxy, "127.0.0.1", 8889);
  QNetworkProxy::setApplicationProxy(proxy);
#else
  if(proxy_prefs.enable_network_proxy)
  {
    QNetworkProxy proxy(proxy_prefs.proxy_type, proxy_prefs.hostname, proxy_prefs.port, proxy_prefs.username, proxy_prefs.password);
    QNetworkProxy::setApplicationProxy(proxy);
  }
  else
  {
    QNetworkProxy::setApplicationProxy(QNetworkProxy(QNetworkProxy::NoProxy));
  }
#endif
}

void cmd_runner::load_state()
{
  qDebug() << "Loading state";

  unpack_mod->load_queue();
  if(zdown_net->get_dl_mgr())
    zdown_net->get_dl_mgr()->invoke_load_state(app_set->fin_dls);

  delete app_set;
  app_set = nullptr;
}

void cmd_runner::load_state_finished(const QString & /*error_text*/)
{
  qDebug() << "Load state finished";

  if(start_download_sched)
    start_download_sched->start();
  else
    QTimer::singleShot(0, this, &cmd_runner::start_downloads);
}

void cmd_runner::start_downloads()
{
  if(zdown_net->get_dl_mgr() == nullptr)
    return;

  if(unpack_mod->settings().unpack_after_download)
  {
    if(unpack_mod->is_queue_empty() == false)
    {
      if(unpack_mod->settings().pause_download_before_unpack == false)
        zdown_net->get_dl_mgr()->invoke_start();

      meta_object_ext::invoke_async(unpack_mod, &unpack::start_unpack_first_item);
      return;
    }
  }

  zdown_net->get_dl_mgr()->invoke_start();

  if(stop_download_sched)
  {
    stop_download_sched->start();
    start_download_sched->deleteLater();
    start_download_sched = nullptr;
  }
}

void cmd_runner::scheduler_stop_triggered()
{
  if(shutdown_system_on_finish)
    schedule_system_shutdown();

  exit_app(false);
}

void cmd_runner::download_finished(const QString & /*error_text*/)
{
  all_downloads_finished = true;

  if(unpack_mod->is_queue_empty() || unpack_mod->has_fatal_error_occured())
    exit_app(true);
  else
    qDebug() << "UNPACK: in progress...";
}

void cmd_runner::downloader_last_item_in_group_finished(const QString & filename)
{
  if(unpack_mod->settings().unpack_after_download)
  {
    if(unpack_mod->settings().pause_download_before_unpack && zdown_net->get_dl_mgr())
      zdown_net->get_dl_mgr()->invoke_stop();

    unpack_mod->add_to_queue(filename);
    meta_object_ext::invoke_async(unpack_mod, &unpack::start_unpack_first_item);
  }
}

void cmd_runner::unpack_queue_finished()
{
  if(all_downloads_finished)
  {
    exit_app(true);
    return;
  }

  if(unpack_mod->settings().pause_download_before_unpack && zdown_net->get_dl_mgr())
    zdown_net->get_dl_mgr()->invoke_start();
}

void cmd_runner::schedule_system_shutdown()
{
  qDebug() << "Scheduling system shutdown";

#ifndef SYSTEM_WINDOWS
  const int exit_code = QProcess::execute("sudo", { "shutdown", "--no-wall", "-P", "+2" });
#else
  const int exit_code = QProcess::execute("shutdown", { "/s", "/t", "120"});
#endif
  qDebug() << "Scheduling system shutdown exit code:" << exit_code;
}

void cmd_runner::exit_app(bool allow_schedule_shutdown)
{  
  if(init_finished == false)
    return;

  init_finished = false; // allow only one exit_app
  qDebug() << "Exiting...";
  unpack_mod->stop_unpack();

  if(shutdown_system_on_finish && all_downloads_finished && allow_schedule_shutdown)
  {
    schedule_system_shutdown();
  }

  QTimer::singleShot(0, this, &cmd_runner::stop_zd_net);
}

void cmd_runner::stop_zd_net()
{
  connect(zdown_net, &zdownloader_net_lib::finished, this, &cmd_runner::zdown_net_finished, Qt::QueuedConnection);
  zdown_net->quit();
  zdown_net->wait();
}

void cmd_runner::zdown_net_finished()
{
  zdown_net->deleteLater();
  QTimer::singleShot(0, qApp, &QCoreApplication::quit);
}

